File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> map = new HashMap<>();
        int maxLen = Integer.MIN_VALUE;
        char[] str = s.toCharArray();
        int left = 0;
        for(int right = 0; right < s.length(); right++){
            char currChar = str[right];

            if(map.containsKey(currChar) && map.get(currChar) >= left){
                
                left = map.get(currChar) + 1;
            }
            
            map.put(currChar,right);
            maxLen = Math.max(maxLen,right - left + 1);
        }

        if(maxLen == Integer.MIN_VALUE) maxLen = 0;

        return maxLen;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        maxFreq = 0
        freq = {}
        result = 0
        left = 0

        for right in range(len(s)):
            freq[s[right]] = freq.get(s[right],0) + 1
            maxFreq = max(freq[s[right]],maxFreq)

            length = right - left + 1

            if (length - maxFreq) > k:
                freq[s[left]] -= 1
                left += 1
            result = max(result, right - left + 1)

        return result
        